### PR TITLE
refactor aws apiGateway event compilation to reduce duplicate code

### PR DESCRIPTION
**Pull request by dougmoscrop**
***Implementing Issue:*** #2025
***Implementing Issue:*** #1738 
***Implementing Issue:*** #2124
***Implementing Issue:*** #2188
***Implementing Issue:*** #2021

## How can we verify it:

`npm test` and integration tests pass for me - if you have more complex services you could test with, that would be great.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

** Splitting up: **

- [x] apiKeys / restApi https://github.com/serverless/serverless/pull/2491
- [ ] authorizers / permissions
- [x] deployment / restApi https://github.com/serverless/serverless/pull/2492
- [x] validate / index https://github.com/serverless/serverless/pull/2496

I will make this PR about the 'methods' and 'resources'

***Is this ready for review?:*** NO

### CODE REVIEW

**Comment by dougmoscrop** [comment]
One big change is using JSON directly instead of JSON.parse/JSON.stringify for the CF resources.

Also looks like one of the tests put an 'empty project' in my main dir, whoops. Will fix that.

**Comment by dougmoscrop** [comment]
LOC went up also because I made the methods.js tests do their own set up each time rather than relying on a global config. It needs to be refined. I thought I saw an issue about it but could not find it just now

**Comment by pmuens** [comment]
DAMN! @dougmoscrop you really rock! Thank you very much for all your hard work you're putting into Serverless! 🌮 🎉

**Comment by dougmoscrop** [comment]
Okay, got it to a net LOC reduction. Have to rebase to pull in the Proxy integration stuff.

**Comment by pouledodue** [comment]
thanks for 1738

**Comment by dougmoscrop** [comment]
@pmuens this is ready for review!

**Comment by pmuens** [comment]
@dougmoscrop thank you for this! I'd like to get it in ASAP as it's a great separation and will reduce lots of pain points we already had / will face if this is not merged.

Could you please rebase once again so that I can do an in-depth testing and merge it soon? Thanks again for your hard work to make Serverless better! 👍 💯

**Comment by dougmoscrop** [comment]
On it. It's a big job each time because Git doesn't understand semantics/refactoring at all, so I basically have to do it by hand.

**Comment by pmuens** [comment]
Great! Thank you very much @dougmoscrop 

Sorry for the inconveniences 😢  😿 

This `method.js` is huge so new functionality for APIG is added there quite frequently. Which resolves in an out of sync PR. But lucky enough you tackled and tamed that beast 🐅 ! Thanks!

**Comment by dougmoscrop** [comment]
Almost done.

**Comment by pmuens** [comment]
Awesome @dougmoscrop! Thank you very much!

**Comment by dougmoscrop** [comment]
@pmuens I think it's good for a review1

File: lib/plugins/aws/deploy/compile/events/apiGateway/index.js [hunk 1]
```diff
@@ -24,37 +27,35 @@ class AwsCompileApigEvents {
       compileRestApi,
       compileApiKeys,
       compileResources,
+      configurePreflightMethods,
```
Comment (dougmoscrop) [inline lib/plugins/aws/deploy/compile/events/apiGateway/index.js]: I think I will rename this to `compileCors` - sorry leftover from pulling in other changes.

**Comment by ac360** [comment]
This is a great clean-up @dougmoscrop.  Thanks for your effort :)

**Comment by pmuens** [comment]
Hey @dougmoscrop thank you very much for your time and hard work on this refactoring.

I just looked through the code and saw that you're using `AWS` as the integration type.
We've introduced `LAMBDA` and `LAMBDA-PROXY` as possible integration types for the events, so the event syntax might look like this:

```
functions:
  foo:
    handler: handler.foo
    events:
      - http:
           path: foo
           method: GET
           integration: LAMBDA
```

This will be translated to the corresponding AWS integration types in the background.

I ran the PR against our integration test suite and unfortunately it broke.

You can run the integration test suite with the following command: `npm run complex-integration-test` (Maybe you want to comment out the non API Gateway tests beforehand in `tests/integration/all.js`).

Do you think it's possible to split this PR into smaller PRs so that they refactor one method at a time. That would make it easier to review and test.

Again. Thank you very much for your help 👍 👍 👍 💯 !
We're almost there and I'd really like to see this refactoring in `master` anytime soon as the unrefactored codebase is hard to understand / work with. Let me know if there's anything I can do tho help you here as you've already spent a tremendous amount of time and work into this PR.

**Comment by dougmoscrop** [comment]
I can work on splitting it up.

As for the integration type: https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js#L584

It's already mapped to AWS/AWS_PROXY, I just moved the mapping to `validate`.

**Comment by pmuens** [comment]
Ah I see!

Thank you for opening up separate PRs. That makes the review process a little bit easier for me :-)
I'll close this PR as we have separate ones.

@dougmoscrop again. You're really killing it 💥 ! Love the work you've done. It's great to have you in the community! 💃 👯 🎉 

/cc @flomotlik
